apps/web/vibes/soul/examples/sections/cart/actions.ts

```diff
@@ -87,3 +91,41 @@ export async function checkoutAction(
 
   return prevState;
 }
+
+export async function couponCodeAction(
+  prevState: Awaited<CouponCodeFormState>,
+  formData: FormData,
+): Promise<CouponCodeFormState> {
+  'use server';
+
+  const submission = parseWithZod(formData, { schema: couponCodeActionFormDataSchema });
+
+  if (submission.status !== 'success') {
+    return {
+      ...prevState,
+      lastResult: submission.reply({ formErrors: ['This code has expired or is invalid!'] }),
```

Comment:
This is an example so it doesn't matter much. But when `submission.status` doesn't equal `'success'`, we can just do `submission.reply()` as the error is a parsing error for one of the fields. Submitting a custom form error would be appropriate when the simulated network request fails, for example.